File: examples/go/yds/function/producer.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/ydb-platform/ydb-go-sdk/v3"
	"github.com/ydb-platform/ydb-go-sdk/v3/topic/topicwriter"
	yc "github.com/ydb-platform/ydb-go-yc"
)

// ProducerHandler handles HTTP requests to write data to YDS stream
func ProducerHandler(w http.ResponseWriter, r *http.Request) {
	ctx := context.Background()

	// Check required environment variables
	ydbEndpoint := os.Getenv("YDB_ENDPOINT")
	topicName := os.Getenv("YDS_TOPIC_ID")
	if ydbEndpoint == "" || topicName == "" {
		http.Error(w, "YDB_ENDPOINT and YDS_TOPIC_ID environment variables must be set", http.StatusInternalServerError)
		return
	}

	// Parse the request
	var req ProducerRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		log.Printf("Error decoding request: %v", err)
		http.Error(w, "Invalid request format", http.StatusBadRequest)
		return
	}

	// Validate request
	if req.Message == "" {
		http.Error(w, "Message is required", http.StatusBadRequest)
		return
	}

	// Create event data
	eventData := map[string]interface{}{
		"message":   req.Message,
		"user_id":   req.UserID,
		"action":    req.Action,
		"timestamp": time.Now().Unix(),
	}

	// Convert to JSON
	jsonData, err := json.Marshal(eventData)
	if err != nil {
		log.Printf("Error marshaling event data: %v", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	// Write to YDS topic
	err = writeToTopic(ctx, ydbEndpoint, topicName, string(jsonData))
	if err != nil {
		log.Printf("Error writing to topic: %v", err)
		http.Error(w, "Failed to write to topic", http.StatusInternalServerError)
		return
	}

	// Return success response
	response := ProducerResponse{
		StatusCode: 200,
		Message:    "Data written to topic successfully",
		StreamID:   topicName,
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(response)
}

// writeToTopic writes data to the YDB topic using the YDB Go SDK
func writeToTopic(ctx context.Context, ydbEndpoint, topicName, data string) error {
	db, err := ydb.Open(ctx, ydbEndpoint,
		yc.WithMetadataCredentials(),
	)
	if err != nil {
		return fmt.Errorf("failed to connect to YDB: %w", err)
	}
	defer db.Close(ctx)

	writer, err := db.Topic().StartWriter(topicName)
	if err != nil {
		return fmt.Errorf("failed to create topic writer: %w", err)
	}
	defer writer.Close(ctx)

	msg := topicwriter.Message{Data: bytes.NewReader([]byte(data))}
	if err := writer.Write(ctx, msg); err != nil {
		return fmt.Errorf("failed to write message to topic: %w", err)
	}

	return nil
}

File: examples/typescript/yds/producer.ts
import { Driver } from '@ydbjs/core';
import { TopicWriter } from '@ydbjs/topic/writer';
import type { Request, Response } from 'express';

// Types matching Go example
interface ProducerRequest {
  message: string;
  user_id: string;
  action: string;
}

interface ProducerResponse {
  status_code: number;
  message: string;
  stream_id?: string;
}

// Main handler function (Express-style)
export async function producerHandler(req: Request, res: Response) {
  const connectionString = process.env['YDB_CONNECTION_STRING'];
  const topicName = process.env['YDS_TOPIC_ID'];
  if (!connectionString || !topicName) {
    res.status(500).json({ status_code: 500, message: 'YDB_CONNECTION_STRING and YDS_TOPIC_ID environment variables must be set' });
    return;
  }

  let body: ProducerRequest;
  try {
    body = req.body;
  } catch (err) {
    res.status(400).json({ status_code: 400, message: 'Invalid request format' });
    return;
  }

  if (!body.message) {
    res.status(400).json({ status_code: 400, message: 'Message is required' });
    return;
  }

  // Create event data
  const eventData = {
    message: body.message,
    user_id: body.user_id,
    action: body.action,
    timestamp: Math.floor(Date.now() / 1000),
  };

  // Write to YDB topic
  try {
    const driver = new Driver(connectionString);
    await driver.ready();
    await writeToTopic(driver, topicName, eventData);
    await driver.destroy();
    const response: ProducerResponse = {
      status_code: 200,
      message: 'Data written to topic successfully',
      stream_id: topicName,
    };
    res.status(200).json(response);
  } catch (err: any) {
    res.status(500).json({ status_code: 500, message: 'Failed to write to topic: ' + err.message });
  }
}

// Helper to write a message to the topic
async function writeToTopic(driver: Driver, topic: string, data: any) {
  const writer = await TopicWriter.create(driver, { topic });
  try {
    const payload = Buffer.from(JSON.stringify(data), 'utf8');
    await writer.write(payload);
  } finally {
    await writer.close();
  }
} 